### ISSUE TITLE
Restrict float attribute values where possible to allow for better xml-validation.

### ISSUE THREAD

**Issue by jukervin** [altoxml/schema#62]
For example ALTO schema allows negative float values in attributes like WIDTH, HEIGHT, HPOS, VPOS where values should be positive.  

PAGE PHYSICAL_IMG_NR

Validating against schema doesn't catch documents where software has created nonsensical values. 

xsd:float is used in following attributes:

PageType
- HEIGHT type="xsd:float" 
- WIDTH type="xsd:float"
- PHYSICAL_IMG_NR" type="xsd:float" : The number of the page within the document.
-> values should be always positive integers?
- ACCURACY type="xsd:float" : Estimated percentage of OCR Accuracy in range from 0 to 100
-> values should be limited to 0-100 floats?

ParagraphStyle
- LEFT type="xsd:float" : Left indent of the paragraph in relation to the column.
- RIGHT type="xsd:float" : Right indent of the paragraph in relation to the column.
- LINESPACE type="xsd:float" :  Line spacing between two lines of the paragraph. Measurement calculated from baseline to baseline.
- FIRSTLINE type="xsd:float" : Indent of the first line of the paragraph if this is different from the other lines. A negative value indicates an indent to the left, a positive value indicates an indent to the right.

BlockType
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"
- ROTATION type="xsd:float" : Tells the rotation of e.g. text or illustration within the block. The value is in degree counterclockwise.
-> Limit to 0-360 float?

SPType
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"

StringType
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"

PageSpaceType
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"

EllipseType
- HPOS type="xsd:float"
- VPOS type="xsd:float"
- HLENGTH type="xsd:float"
- VLENGTH type="xsd:float"
- ROTATION type="xsd:float"
An ellipse shape. HPOS and VPOS describe the center of the ellipse. HLENGTH and VLENGTH are the width and height of the described ellipse.
The attribute ROTATION tells the rotation of the e.g. text or illustration within the block. The value is in degrees counterclockwise.
-> Limit to 0-360 float?

CircleType
A circle shape. HPOS and VPOS describe the center of the circle.
- HPOS type="xsd:float"
- VPOS type="xsd:float"
- RADIUS type="xsd:float"

formattingAttributeGroup
- FONTSIZE type="xsd:float" : The font size, in points (1/72 of an inch).

HYP
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"

TextLine
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"
- BASELINE type="xsd:float"

GlyphType
- HEIGHT type="xsd:float"
- WIDTH type="xsd:float"
- HPOS type="xsd:float"
- VPOS type="xsd:float"

**Comment by jukervin** [comment]
Based on the discussion in the meeting 2020-12-13 the ROTATION can have valid negative values.

**Comment by jukervin** [comment]
Limiting values to positive will break backward compatibility so this can be changed in ALTO 5.0 release at the earliest.

**Comment by Ra1phM** [comment]
I agree that limiting WIDTH, HEIGHT, HPOS, VPOS to positive values would make sense. From my understanding, HPOS and VPOS  are always in relation to the entire page, so values outside of the page's real dimensions (e.g. (-100, -100)) should not exist.

However, for the ParagraphStyle LEFT and RIGHT indent, I am not sure if positive values should be enforced, because the indent value is relative to the paragraph and even if it is not a good practice, it would still be valid in the same way that negative values (or positions, margins and paddings) in HTML are accepted.

**Comment by jukervin** [comment]
Maybe XSD 1.1 schema asserts  could be used to create validations that rely on other element values: PrintSpace can't be larger than Page etc.

**Comment by cipriandinu** [comment]
A new branch added (issue-62) to make changes for this topic. There are several things to discuss, since the only option to completely implement this is to use xsd 1.1 (to restrict not only some values to be positive, but also restrictions like height of a block + vpos < page height). On the new branch, there are just fie restrictions on Page level, but there is much more to be done.

One topic to clarify is if we want to go with this solution and enforce validation with xsd 1.1 processors. Or if we implement only simple restrictions (positive values, no relative restrictions)? If we go for full solution, probably when switching to 5.0 would be a good moment
